mapper/orm/src/main/java/org/hibernate/search/mapper/orm/batchindexing/impl/MassIndexerImpl.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.orm.batchindexing.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import org.hibernate.CacheMode;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.search.mapper.orm.MassIndexer;
+import org.hibernate.search.mapper.orm.batchindexing.spi.MassIndexerWithTenant;
+import org.hibernate.search.mapper.orm.impl.HibernateSearchContextService;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.orm.mapping.spi.HibernateOrmMapping;
+import org.hibernate.search.mapper.pojo.work.spi.PojoWorkPlan;
+import org.hibernate.search.util.impl.common.Executors;
+import org.hibernate.search.util.impl.common.LoggerFactory;
+import org.hibernate.search.util.impl.common.StringHelper;
+
+/**
+ * Prepares and configures a BatchIndexingWorkspace to start rebuilding
+ * the indexes for all entity instances in the database.
+ * The type of these entities is either all indexed entities or a
+ * subset, always including all subtypes.
+ *
+ * @author Sanne Grinovero
+ */
+public class MassIndexerImpl implements MassIndexerWithTenant {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final PojoWorkPlan workPlan;
+	private final HibernateOrmMapping mapping;
+	private final SessionFactoryImplementor sessionFactory;
+
+	protected final Set<Class<?>> rootEntities;
+
+	// default settings defined here:
+	private int typesToIndexInParallel = 1;
+	private int documentBuilderThreads = 6; //loading the main entity, also responsible for loading of lazy @IndexedEmbedded collections
+	private int objectLoadingBatchSize = 10;
+	private long objectsLimit = 0; //means no limit at all
+	private CacheMode cacheMode = CacheMode.IGNORE;
+	private boolean optimizeAtEnd = true;
+	private boolean purgeAtStart = true;
+	private boolean optimizeAfterPurge = true;
+	//TODO: restore the MassIndexerProgressMonitor
+	private int idFetchSize = 100; //reasonable default as we only load IDs
+	private String tenantIdentifier;
+	private Integer idLoadingTransactionTimeout;
+
+	protected MassIndexerImpl(SessionFactoryImplementor sessionFactory, Class<?>... entities) {
+		this.sessionFactory = sessionFactory;
+		this.mapping = sessionFactory.getServiceRegistry().getService( HibernateSearchContextService.class ).getMapping();
+		this.workPlan = mapping.createSearchManager( sessionFactory.createEntityManager() ).createWorkPlan();
+		this.rootEntities = toRootEntities( entities );
```

Comment:
Please don't remove any code without at least leaving a TODO for later. You removed this:

```
		if ( extendedIntegrator.isJMXEnabled() ) {
			monitor = new JMXRegistrar.IndexingProgressMonitor();
		}
		else {
			monitor = new SimpleIndexingProgressMonitor();
		}
```